### PR TITLE
build/ops: rpm: apply epoch only if %epoch macro is defined

### PR DESCRIPTION
**Pull request by smithfarm**
This allows SUSE to drop a maintenance-intensive downstream patch.

### CODE REVIEW

**Comment by smithfarm** [comment]
https://shaman.ceph.com/builds/ceph/wip-cond-epoch/

File: ceph.spec.in [hunk 1]
```diff
@@ -82,10 +84,10 @@ ExclusiveArch:  x86_64 aarch64
 #################################################################################
 # dependencies that apply across all distro families
 #################################################################################
-Requires:       ceph-osd = %{epoch}:%{version}-%{release}
-Requires:       ceph-mds = %{epoch}:%{version}-%{release}
-Requires:       ceph-mgr = %{epoch}:%{version}-%{release}
-Requires:       ceph-mon = %{epoch}:%{version}-%{release}
+Requires:       ceph-osd = %{?epoch:%{epoch}:}%{version}-%{release}
```
Comment (tchaikov) [inline ceph.spec.in]: can we define a variable for epoch like
```
%define _epoch %{?epoch:%{epoch}:}
```
and use it afterwards ?

Comment (smithfarm) [inline ceph.spec.in]: I think we can, and `_epoch` is a good name for it. I just don't know whether to use `%define` or `%global` ?

Comment (ktdreyer) [inline ceph.spec.in]: `%global` sounds good to me

Comment (tchaikov) [inline ceph.spec.in]: okay, i was just googling for how to define a variable in rpm script, and `%define` popped out. if you guys think `%global` is better, i am all for it! =)

**Comment by smithfarm** [comment]
Thanks for looking; modified and repushed.

**Comment by smithfarm** [comment]
https://fedoraproject.org/wiki/PackagingDrafts/global_preferred_over_define

**Comment by tchaikov** [comment]
@smithfarm thanks for the link!

**Comment by smithfarm** [comment]
I will install the packages in CentOS and SLE VMs and check the dependencies for epoch sanity.

**Comment by smithfarm** [comment]
https://shaman.ceph.com/builds/ceph/wip-cond-epoch/ce5d6818e8f2551d8f8b155c372eabcd64146dcc/

File: ceph.spec.in [hunk 2]
```diff
@@ -58,13 +58,19 @@
...
+
 #################################################################################
 # main package definition
 #################################################################################
 Name:		ceph
 Version:	@VERSION@
 Release:	@RPM_RELEASE@%{?dist}
+%if 0%{?fedora} || 0%{?rhel}
 Epoch:		1
+%endif
```
Comment (tchaikov) [inline ceph.spec.in]: https://docs.fedoraproject.org/ro/Fedora_Draft_Documentation/0.1/html/RPM_Guide/ch09s03.html

> Warning
Avoid using the Epoch: directive if at all possible. It is far better to use a sane version-numbering scheme than to try to resolve the mess with epoch values. The main problems with using an epoch value are that epochs are hidden from users in most cases, and using epochs can lead to very strange-looking tasks such as a newer package with a version number that looks older than the older package.

@ktdreyer maybe we should avoid using `Epoch`?

Comment (smithfarm) [inline ceph.spec.in]: @tchaikov Apparently, once the Epoch is applied, there is no way to undo it.

Comment (tchaikov) [inline ceph.spec.in]: @smithfarm thanks! got it, no U turn then =(

Comment (ktdreyer) [inline ceph.spec.in]: Right :(

**Comment by b-ranto** [comment]
FYI: This broke package installation on centos/rhel, yum can no longer resolve the dependencies because the versions do not match. My first guess is that we are missing `:` character in the versions.

**Comment by b-ranto** [comment]
Anyway, I am not sure what is the issue here? Why don't you just define epoch to 0 (the default epoch) if you do not want to bump it? The rpm treat "no epoch" the same way as epoch 0. If you did not want to bump the epoch, you could just conditionally make it 0.

**Comment by b-ranto** [comment]
btw: We (well I) originally introduced Epoch because somebody pushed unstable version of ceph into epel and that overrode the packages provided by upstream. At that point, that was pretty much the only way to solve the problem in a sane way (yum-plugin-priorities were still broken, ...).

**Comment by smithfarm** [comment]
Yeah, I was going to test it on CentOS first, before merging :-(

**Comment by smithfarm** [comment]
Maybe the `%global` is too far up in the file? Opening another PR to address.

**Comment by b-ranto** [comment]
Thanks, that other PR helped and I can now install the packages.

**Comment by smithfarm** [comment]
> Thanks, that other PR helped and I can now install the packages.

Yeah, I was sweating bullets early this morning :-) Good to know it's fixed.

**Comment by smithfarm** [comment]
> Anyway, I am not sure what is the issue here? Why don't you just define epoch to 0 (the default epoch) if you do not want to bump it?

I believe I tried explicitly setting `Epoch: 0` and OBS refused to build it. SUSE does not allow epochs in the spec file at all.

**Pull request merged**